Login view

mammadu opened the pull request:
Create login page
Create user page

mammadu committed:
removed unimplemented tests

mammadu committed:
Create login field in login/new.html.erb

mammadu committed:
update login form label

mammadu committed:
Login now creates a session

mammadu committed:
login controller redirects to userpage upon login

mammadu committed:
changed params key to address, email address should now be sent between user and login controllers

mammadu committed:
update login controller test
'creates a session' test corrected so that params key is 'address'

mammadu committed:
create private function 'is_user' to search database by email address

mammadu committed:
refactor login controller and login controller spec
all tests pass
created before and after functions for testing
refactored loginCreate to use an if/else statement

mammadu committed:
login page shows message if login information is incorrect

mammadu committed:
Create button to create account

mammadu committed:
add validations for user model

mammadu committed:
add validation for email uniqueness

mammadu committed:
working on refactoring login_controller to move is_user to the user model

mammadu committed:
create form to submit new account info

mammadu committed:
can now create users in database using new form and user controller

mammadu committed:
fix rubocop offenses

mammadu committed:
Fix rubocop violations. Add unique index to email column

Dan-Burnette commented on app/models/user.rb:
```diff
@@ -1,4 +1,10 @@
+  validates :name,  :username, :email, presence: true
+  validates :email, uniqueness: true
+
+  def user?(email_address)
+    User.find_by email: email_address
+  end
```
Question mark at end of method should only be used when it definitely will return either `true` or `false`. 

Another developer may be confused by this because it actually returns a `User` object or `nil`

mammadu replied:
Good point, change implemented

Dan-Burnette commented on app/controllers/login_controller.rb:
```diff
@@ -1,5 +1,18 @@
+  def create
+    email_address = params['email']
+    user = User.new
+    if user.user?(email_address)
```
`user.user?` is kinda confusing. It reads like asking a user whether they are a user, which is just a bit funky if you ask me.

It's very typical in controllers to just perform the lookup on the main object the controller action interacts with, no need to hide that away in another class.  It's also common to inline the params into the query.

For instance:

`if user.user?(email_address)` could be rewritten as `if User.find_by(email: params['email']`)  or even:
```
user = User.find_by(email: params['email'])
if user # truthy if a user object was found, falsy if nil
... do stuff ...
```

Also params typically are referenced as symbols, so it might be `params[:email]` instead

Dan-Burnette replied:
Totally different point: it looks like in your current setup you basically ran `user = User.new` just so that you can then access the method on the user instance: `user.user?`. 

In cases like this, a class method might make more sense.

E.g.

```
class User < ApplicationRecord
  validates :name,  :username, :email, presence: true
  validates :email, uniqueness: true

  def user?(email_address)
    User.find_by email: email_address
  end
end

# user = User.new
# user.user?(email)

```

Becomes 
```
class User < ApplicationRecord
  validates :name,  :username, :email, presence: true
  validates :email, uniqueness: true

  def self.user?(email_address)
    User.find_by email: email_address
  end
end

# used like:
# User.user?(email_address)
```

That way you don't need to instantiate an object via new to access the method

mammadu replied:
Thanks for highlighting how to use 'self' - I did not know how to call class methods without instatiating an object

I changed ```user?``` to ```get_user```

mammadu committed:
add email validations
user model now has email validations
add email_validator gem
clarify user account creation errors

mammadu committed:
rename user? to get_user for clarity
